internal/security/redis/main.go

```diff
@@ -0,0 +1,85 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * @author: Andre Srinivasan
+ *******************************************************************************/
+
+package redis
+
+import (
+	"context"
+	"os"
+
+	"github.com/edgexfoundry/edgex-go/internal"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/bootstrap/handlers/database"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/telemetry"
+	"github.com/edgexfoundry/edgex-go/internal/security/redis/config"
+	"github.com/edgexfoundry/edgex-go/internal/security/redis/container"
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap"
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/flags"
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/handlers/httpserver"
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/handlers/secret"
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/interfaces"
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/startup"
+	"github.com/edgexfoundry/go-mod-bootstrap/di"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+	"github.com/gorilla/mux"
+)
+
+func Main(ctx context.Context, cancel context.CancelFunc, router *mux.Router, readyStream chan<- bool) {
+	startupTimer := startup.NewStartUpTimer(clients.SecurityBootstrapRedisKey)
+
+	// All common command-line flags have been moved to DefaultCommonFlags.
+	f := flags.New()
+	f.Parse(os.Args[1:])
+
+	configuration := &config.ConfigurationStruct{}
+	dic := di.NewContainer(di.ServiceConstructorMap{
+		container.ConfigurationName: func(get di.Get) interface{} {
+			return configuration
+		},
+	})
+
+	httpServer := httpserver.NewBootstrap(router, false) // Not a real HTTP server; false == don't listen/serve
+
+	// bootstrap.Run is needed for the underlying configuration system. Conveniently, it also creates
+	// a pipeline of functions as the list of BootstrapHandler's is executed in order. Note that
+	// bootstrap.Run will return when the last thread spawned by the the BoostrapHandler's returns. In
+	// this case, telemetry.BootstrapHandler creates a thread that effectively will never return
+	// therefore bootstrap.Run will never return.
+	//
+	// httpServer.BootstrapHandler is needed by database.BootstrapHandler.
+	//
+	// secret.BootstrapHandler connects to Vault.
+	//
+	// database.BootstrapHandler (the first time) sets up the Redis connection. This works because
+	// config.GetDatabaseInfo checks if Redis has been secured.
+	bootstrap.Run(
+		ctx,
+		cancel,
+		f,
+		clients.SecurityBootstrapRedisKey,
+		internal.ConfigStemCore+internal.ConfigMajorVersion,
+		configuration,
+		startupTimer,
+		dic,
+		[]interfaces.BootstrapHandler{
+			telemetry.BootstrapHandler,
+			httpServer.BootstrapHandler, // This is a fake to keep NewDatabase happy
+			secret.NewSecret().BootstrapHandler,
+			noAuthConnect,
+			credentialSet,
+			database.NewDatabase(httpServer, configuration).BootstrapHandler,
```

Comment:
Why needed when you are already creating the DBClient in noAuthConnect. This needs to be reworked to have only one handler that does noAuthConnect & credentialSet that way if credentials already set, it just skips attempting to set them again and goes to sleep.

Reply:
Done